File: Python/ByTier/E/1810_E_Monsters.py
# Problem: CF 1810 E - Monsters
# https://codeforces.com/contest/1810/problem/E

"""
Algorithm: Union Find + Sorting
Time Complexity: O(n * log(n) + m * alpha(n)) where alpha(n) is inverse Ackermann function
Space Complexity: O(n + m)

The problem involves determining if we can defeat all monsters in a graph. 
Each monster has a danger value 'a[i]' which means we must defeat at least a[i] other monsters before defeating it.
We can only move through edges, and we can only move to a monster if we've already defeated it or can defeat it now.
This is solved by:
1. Sorting vertices by their danger values
2. For each vertex in sorted order, we check if it's reachable from a previously defeated monster
3. We use UnionFind to track connected components and whether we can reach each component
4. We mark if a vertex is reachable based on whether its components have sufficient defeated monsters

The key insight is that monsters can be defeated in order of their danger values, and we use UnionFind to group connected vertices.
"""

from sys import stdin

inf = 1 << 32


class FastIO:
    def __init__(self):
        self.random_seed = 0
        return

    @staticmethod
    def read_int():
        return int(stdin.readline().rstrip())

    @staticmethod
    def read_float():
        return float(stdin.readline().rstrip())

    @staticmethod
    def read_list_ints():
        return list(map(int, stdin.readline().rstrip().split()))

    @staticmethod
    def read_list_ints_minus_one():
        return list(map(lambda x: int(x) - 1, stdin.readline().rstrip().split()))

    @staticmethod
    def read_str():
        return stdin.readline().rstrip()

    @staticmethod
    def read_list_strs():
        return stdin.readline().rstrip().split()

    def get_random_seed(self):
        import random

        self.random_seed = random.randint(0, 10**9 + 7)
        return

    @staticmethod
    def st(x):
        return print(x)

    @staticmethod
    def lst(x):
        return print(*x)

    @staticmethod
    def max(a, b):
        return a if a > b else b

    @staticmethod
    def min(a, b):
        return a if a < b else b

    @staticmethod
    def ceil(a, b):
        return a // b + int(a % b != 0)

    @staticmethod
    def accumulate(nums):
        n = len(nums)
        pre = [0] * (n + 1)
        for i in range(n):
            pre[i + 1] = pre[i] + nums[i]
        return pre


class UnionFind:
    def __init__(self, n: int) -> None:
        self.root_or_size = [-1] * n
        self.part = n
        return

    def find(self, x):
        y = x
        while self.root_or_size[x] >= 0:

            x = self.root_or_size[x]
        while y != x:
            self.root_or_size[y], y = x, self.root_or_size[y]
        return x

    def union(self, x, y):
        root_x = self.find(x)
        root_y = self.find(y)
        if root_x == root_y:
            return False
        if self.root_or_size[root_x] < self.root_or_size[root_y]:
            root_x, root_y = root_y, root_x
        self.root_or_size[root_y] += self.root_or_size[root_x]
        self.root_or_size[root_x] = root_y
        self.part -= 1
        return True

    def union_left(self, x, y):
        root_x = self.find(x)
        root_y = self.find(y)
        if root_x == root_y:
            return False
        self.root_or_size[root_x] += self.root_or_size[root_y]
        self.root_or_size[root_y] = root_x
        self.part -= 1
        return True

    def union_right(self, x, y):
        root_x = self.find(x)
        root_y = self.find(y)
        if root_x == root_y:
            return False
        self.root_or_size[root_y] += self.root_or_size[root_x]
        self.root_or_size[root_x] = root_y
        self.part -= 1
        return True

    def get_node_size(self, x):
        return -self.root_or_size[self.find(x)]


class Solution:
    def __init__(self):
        return

    @staticmethod
    def main(ac=FastIO()):
        for _ in range(ac.read_int()):
            n, m = ac.read_list_ints()
            nums = ac.read_list_ints()
            # Create a unique value for each node combining danger with index
            nums = [num * n + i for i, num in enumerate(nums)]

            edge = [[] for _ in range(n)]
            for _ in range(m):
                u, v = ac.read_list_ints_minus_one()
                # Ensure u has greater or equal danger value than v for processing
                if nums[u] < nums[v]:
                    u, v = v, u
                edge[u].append(v)

            reach = [0] * n
            uf = UnionFind(n)
            # Sort nodes by their danger values (and index)
            nums.sort()
            for val in nums:
                num, i = val // n, val % n
                # If danger level is 0, it's reachable
                if not num:
                    reach[i] = 1
                # Check each neighbor for connection
                for j in edge[i]:
                    # If neighbor was visited and we have sufficient defeated monsters, mark current as reachable
                    if reach[uf.find(j)] and num <= uf.get_node_size(j):
                        reach[i] = 1
                    # Union the current and neighbor vertices
                    uf.union_left(i, j)
            # Check if all nodes are in one component and the initial node is reachable
            ac.st("YES" if uf.part == 1 and reach[uf.find(0)] else "NO")
        return


Solution().main()


# https://github.com/VaHiX/CodeForces/

File: CPP/ByRound/1810/1810_E_Monsters.cpp
// Problem: CF 1810 E - Monsters
// https://codeforces.com/contest/1810/problem/E

/*
Algorithm: Union-Find (Disjoint Set Union) with sorting by monster danger
Purpose: Determine if all monsters in a graph can be defeated, starting from a
vertex with danger 0, moving through edges only when the monster's danger
requirement is met.

Approach:
1. For each vertex, store its monster danger value.
2. Use Union-Find (DSU) to merge connected components, considering the danger
values.
3. Sort edges by the maximum danger of the two vertices they connect.
4. Process edges in increasing order of max danger, merging components and
updating validity.
5. Check if any component reachable from a start vertex (with danger 0) can
contain all vertices.

Time Complexity: O(m log m + n α(n)) where α is the inverse Ackermann function.
Space Complexity: O(n + m)

The solution works because:
- Edges are processed in order of increasing danger, so we can safely merge
components.
- For each component, we track whether it's valid (contains a start vertex with
danger 0 and sufficient monsters defeated).
- At the end, we check if the component containing vertex 1 is valid and
includes all vertices.
*/

#include <algorithm>
#include <iostream>
#include <stddef.h>
#include <utility>

using namespace std;
const int maxn = 2e5 + 5;
pair<int, int> edges[maxn];
int a[maxn];
int root[maxn];
bool is_valid[maxn];
int get_root(int u) {
  while (root[u] > 0)
    u = root[u];
  return u;
}
void join(int u, int v) {
  int cost = max(a[u], a[v]);
  u = get_root(u);
  v = get_root(v);
  if (u == v)
    return;
  if (-root[v] > -root[u]) {
    swap(u, v);
  }
  int new_valid =
      (is_valid[u] && -root[u] >= cost) || (is_valid[v] && -root[v] >= cost);
  root[u] += root[v];
  root[v] = u;
  is_valid[u] = new_valid;
}
int main() {
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);
  int t;
  cin >> t;
  while (t--) {
    int n, m;
    cin >> n >> m;
    for (int i = 1; i <= n; ++i) {
      cin >> a[i];
      if (a[i] == 0) {
        is_valid[i] = 1;
      } else
        is_valid[i] = 0;
      root[i] = -1;
    }
    for (int i = 1; i <= m; ++i) {
      cin >> edges[i].first >> edges[i].second;
    }
    sort(edges + 1, edges + m + 1, [&](auto &e1, auto &e2) {
      int costE1 = max(a[e1.first], a[e1.second]);
      int costE2 = max(a[e2.first], a[e2.second]);
      return costE1 < costE2;
    });
    for (int i = 1; i <= m; ++i) {
      const auto [u, v] = edges[i];
      join(u, v);
    };
    if (is_valid[get_root(1)] && -root[get_root(1)] == n) {
      cout << "YES\n";
    } else {
      cout << "NO\n";
    }
  }
  return 0;
}

// https://github.com/VaHiX/CodeForces/